### PR TITLE
[fix] Treat clang-diagnostic-* checkers as compiler flags

### PR DESCRIPTION
**Pull request by vodorok**
Previously the handling of clang-diagnostics was flawed because tidy cannot enable compiler warnings when specified  as a checker, only suppression of the resulting reports are available:

From the official [clang-tidy docs](https://clang.llvm.org/extra/clang-tidy/):
 > Clang diagnostics are treated in a similar way as check diagnostics. Clang diagnostics are displayed by clang-tidy and can be filtered out using the -checks= option. However, the -checks= option does not affect compilation arguments, so it cannot turn on Clang warnings which are not already turned on in the build configuration. The -warnings-as-errors= option upgrades any warnings emitted under the -checks= flag to errors (but it does not enable any checks itself).

After this patch all of these checkers will be treated as compiler warning flags, this way the compiler warnings can be enabled as clang-diagnostics.
There is an added benefit of this method, as the checker prefix based activation is also working now.

Before the patch:
* Trying to use `-e clang-diagnostic-...` notation did not produce any result because tidy wouldn't insert the corresponding compiler warning to the input compilation command. We thought adding it to the checks list enabled the warning, but it didn't.

After the patch:
* Using `-e W...` notation results in a warning from CodeChecker to use the clang-diagnostic-unused-variable counterpart.
* Using the `-e clang-diagnostic-...` now properly works. and enables the related compiler warnings.
* When using `--enable-all`, instead of adding every compiler warning individually to the build command, `-Weverything` is added.

There are some considerations regarding the old `-W...` notation.
* Using the `-W...` notation is be deprecated, but for the time being still usable.\
* With `--enable-all`  the usage of `-d -W...` notation is ignored, and results in a warning. The `-d compiler-diagnostics-...` is the recommended way to disable a compiler warning.

### CODE REVIEW

File: analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py
Comment (Szelethus) [inline analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py:232]: I feel like this branch should be moved to `get_compiler_warning_name`, that function is the authority on what we interpret to be a compiler warning.

File: analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py
Comment (Szelethus) [inline analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py:236]: This feels a little sterile. How about "Since 6.22.0, using 'W' prefix for compiler warnings 
 is deprecated. Please change from '`{checker_name}`' to 'clang-diagnostic-`{checker_name[1:]}'`."

File: analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py
Comment (Szelethus) [inline analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py:249]: We already have the variable `warning_name`, that holds the information on whether we recognized the checker as compiler warning, no? Why are we reinveting the wheel instead of reusing the result of `get_compiler_warning_name`?

**Review by Szelethus — CHANGES_REQUESTED** [review]
The story here, as I understand it is the following. Both the CLI and the GUI could incorrectly let you believe that we handle compiler warnings as checkers, but that is only party true. We can't reliably enable/disable them. Not only that, the checker name they are displayed with start with "compiler-diagnostic-<warning name>", but they must be named as "W<warning name>" on the CLI.

> Previously the handling of clang-diagnostics was flawed, because tidy cannot enable those as a checker, only suppression of the resulting reports are available:

Yes, because compiler warnings are flags to the _compiler_, not the _analyzer_ component. This makes some sense for a compiler developer, but as a user of a static analyzer tool, compiler warnings _should_ be checkers. If clang-tidy won't do it on its own for whetever reason, we need to make it so that CodeChecker acts as if they were.

> https://github.com/kimgr/clang-tools-extra/blob/master/docs/clang-tidy.rst

This is painfully out of date. Back in the day, LLVM didn't have a single monorepo housing all of its subprojects, but all subprojects had their own repository (imagine how painful an LLVM-wide refactoring used to be!). Not only is this repo from that era, it is ~6 years before the merging, and its not even the official mirror either.

This is probably what you are after:
https://clang.llvm.org/extra/clang-tidy/#suppressing-undesired-diagnostics

> After this patch all of these checkers will be treated as compiler warning flags,

Not the other way around?

> this way the compiler warnings can be enabled as clang-diagnostics.

Meaning that the name of the checker on the GUI and the name to be used on the CLI will be identical.

**Comment by vodorok** [comment]
@Szelethus 
> The story here, as I understand it is the following. Both the CLI and the GUI could incorrectly let you believe that we handle compiler warnings as checkers, but that is only party true. We can't reliably enable/disable them. Not only that, the checker name they are displayed with start with "compiler-diagnostic-", but they must be named as "W" on the CLI.

I think you are mistaken here. The intent is to forgo the `W` designation, and from the perspective of the CodeChecker CLI use only the `clang-diagnostic` naming.

The following matrix shows the current state of compiler warning activation

|         a         | -W...           | -Wno...  | clang-diagnostics... |
| ------------- |-------------| -----|--------|
| --enable      | adds -W... flag to the build command | adds -Wno... flag to the build command | adds clang-diagnostic-... to the checks list / not working |
| --disable     | adds -Wno... flag to the build command  | adds -W.. flag to the build command | adds -clang-diagnostic-.. to the checks list / suppresses the compiler warnings based on checker name prefixes |

> > Previously the handling of clang-diagnostics was flawed, because tidy cannot enable those as a checker, only suppression of the resulting reports are available:
> 
> Yes, because compiler warnings are flags to the _compiler_, not the _analyzer_ component. This makes some sense for a compiler developer, but as a user of a static analyzer tool, compiler warnings _should_ be checkers. If clang-tidy won't do it on its own for whetever reason, we need to make it so that CodeChecker acts as if they were.

This is the intention of the patch.

> > https://github.com/kimgr/clang-tools-extra/blob/master/docs/clang-tidy.rst
> 
> This is painfully out of date. Back in the day, LLVM didn't have a single monorepo housing all of its subprojects, but all subprojects had their own repository (imagine how painful an LLVM-wide refactoring used to be!). Not only is this repo from that era, it is ~6 years before the merging, and its not even the official mirror either.

Thanks for pointing out that I included an outdated (and forked) documentation. The official docs still state the same.
https://clang.llvm.org/extra/clang-tidy/
Look below the checker group table


From the official clang-tidy docs:
> Clang diagnostics are treated in a similar way as check diagnostics. Clang diagnostics are displayed by clang-tidy and can be filtered out using the -checks= option. However, the -checks= option does not affect compilation arguments, so it cannot turn on Clang warnings which are not already turned on in the build configuration. The -warnings-as-errors= option upgrades any warnings emitted under the -checks= flag to errors (but it does not enable any checks itself).

> This is probably what you are after: https://clang.llvm.org/extra/clang-tidy/#suppressing-undesired-diagnostics
This was it thanks!

> > After this patch all of these checkers will be treated as compiler warning flags,
> 
> Not the other way around?

No, because of the previously mentioned facts.

> > this way the compiler warnings can be enabled as clang-diagnostics.
> 
> Meaning that the name of the checker on the GUI and the name to be used on the CLI will be identical.
This is true.

Before the patch: 
* You tried to use `-e clang-diagnostic-vla` you did not get any result because tidy wouldn't insert the corresponding compiler warning to the input compilation command.

After the patch:
* You use `-e Wunused-variable` you get a warning from CodeChecker to use the `clang-diagnostic-unused-variable` counterpart.
* You can use the `clang-diagnostic-vla` naming to enable the vla-related compiler warnings.

File: analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py
Comment (vodorok) [reply analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py:236]: Good suggestion, thanks.

File: analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py
Comment (vodorok) [reply analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py:232]: I thought about this, but `get_compiler_warning_name` is not a private member. In the future somebody could call it in some other context, then it would warning even if it is not related.

File: analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py
Comment (Szelethus) [reply analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py:252]: This english is still a little broken.

File: analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py
Comment (Szelethus) [inline analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py:349]: This deserves some explanation in the code. Possibly in the summary.

**Review by bruntib — APPROVED** [review]

File: web/tests/functional/report_viewer_api/test_get_run_results.py
Comment (Szelethus) [inline web/tests/functional/report_viewer_api/test_get_run_results.py:271]: Do you suspect where this needs to be fixed in the code? Maybe you could mention that ("I suspect that `worst_sorterino` needs to be tinkered with in `analyze.py`").

**Review by Szelethus — CHANGES_REQUESTED** [review]
Tiny things left to attend to.

File: analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py
Comment (vodorok) [reply analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py:349]: Yes this is a new development, I will include the explanation in the summary.

File: web/tests/functional/report_viewer_api/test_get_run_results.py
Comment (vodorok) [reply web/tests/functional/report_viewer_api/test_get_run_results.py:271]: I have some rough ideas, will include these in the comments, unfortunately, I don't have any conclusion yet.
I will create an issue too.

File: analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py
Comment (Szelethus) [reply analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py:255]: I still think this is bad design. The code would look more readable if we moved that logic to `get_compiler_warning_name` (where it belongs), and we really need to, check whether this is a disabled clang-diagnostic _after the fact_.

**Review by Szelethus — CHANGES_REQUESTED** [review]
Ugh, sorry :)

The summary is nice.

**Review by Szelethus — APPROVED** [review]
LGTM! If you wanna make things a bit simpler, go ahead, I'll leave that to you.

**Pull request merged**
